src/wpublish/QueryStatistics.h

```diff
@@ -0,0 +1,173 @@
+    /// when they finish.
+    /// Copying is fine, but setting one equal to another would cause issues.
+    class SchedulerTasksInfo {
+    public:
+        SchedulerTasksInfo() = delete;
+        SchedulerTasksInfo(std::shared_ptr<wsched::SchedulerBase> const& sched_, int taskCount_)
+            : scheduler(sched_), schedulerName(sched_->getName()), firstTaskAdded(CLOCK::now()),
+              mostRecentTaskAdded(firstTaskAdded), taskCount(taskCount_) {}
+        SchedulerTasksInfo(SchedulerTasksInfo const&) = default;
+        SchedulerTasksInfo& operator=(SchedulerTasksInfo const&) = delete;
+
+        std::weak_ptr<wsched::SchedulerBase> const scheduler;
+        std::string const schedulerName;
+        TIMEPOINT const firstTaskAdded; ///< The time the first `task` was added to this scheduler.
+        TIMEPOINT mostRecentTaskAdded; ///< The time the last `task` was added to this scheduler.
+        int taskCount = 0; ///< how many tasks were added to this scheduler.
+    };
+
+    /// Type definition for consistency.
+    using SchedTasksInfoMap = std::map<std::string, SchedulerTasksInfo>;
+
+    /// Track that `numberOfTasksAdded` has been added to `sched` where all of the `Task`s are related
+    /// to this UserQuery.
+    void tasksAddedToScheduler(std::shared_ptr<wsched::SchedulerBase> const& sched, int numberOfTasksAdded);
+
+    /// Return a copy of `_taskSchedMap`.
+    SchedTasksInfoMap getSchedulerTasksInfoMap();
+
+    TIMEPOINT const creationTime;
+    QueryId const queryId;
+
+    /// Return a json object containing high level data, such as histograms.
+    nlohmann::json getJsonHist() const;
+
+    /// Retrieve a status of the tasks as defined by the query selector.
+    /// @note If no restrictors are specified in the task selector then the method
+    /// can return a very large object. So it should be used sparingly.
+    /// @param taskSelector Task selection criterias.
+    /// @return a json object containing information about tasks in the requested scope.
+    nlohmann::json getJsonTasks(wbase::TaskSelector const& taskSelector) const;
+
+    friend class QueriesAndChunks;
+    friend std::ostream& operator<<(std::ostream& os, QueryStatistics const& q);
+
+private:
+    explicit QueryStatistics(QueryId const& queryId);
+    bool _isMostlyDead() const;
+
+    mutable std::mutex _qStatsMtx;
+
+    std::chrono::system_clock::time_point _touched = std::chrono::system_clock::now();
+
+    int _size = 0;
+    int _tasksCompleted = 0;
+    int _tasksRunning = 0;
+    int _tasksBooted = 0;                   ///< Number of Tasks booted for being too slow.
+    std::atomic<bool> _queryBooted{false};  ///< True when the entire query booted.
+
+    double _totalTimeMinutes = 0.0;
+
+    std::vector<wbase::Task::Ptr> _tasks;  ///< A collection of all tasks of the query
+
+    util::Histogram::Ptr _histTimeRunningPerTask;  ///< Histogram of SQL query run times.
```

Comment:
A few comments on the histogram data members. Firstly, I'm not seeing any _include_ statements in this header for those. Since you depend on the typedef for `::Ptr` then I would expect something like:
```
#include "util/Histogram.h"
```
A reason why this code might still compile could be purely accidental - because the header for `wpublish/QservStatistics.h` is used in some context where the `Histogram` is already included. We can't rely on this. Normally, all data types mentioned in a particular C++ header must be fully declared (if used by value, etc.) or forward-declared (if there are pointers or references to that type).

My second comment (suggestion) is directly related to the last statement. The `::Ptr` is "evil" in this case as it forces you to do the full include of the relevant header:
```
#include "util/Histogram.h"

```
Where you could simply do the light-weight forward declaration:
```
namespace lsst::qserv::util {
class Histogram;
}
```
And then do:
```
std::shared_ptr<util::Histogram> _histTimeRunningPerTask;
```
The lightweight forward declarations are helpful as they are known to speed up the compilation of a project.

It's okay to use `::Ptr` in the implementation (`*.cc`) files where the corresponding type is used.

Reply:
The verbose explanation is a bit overkill. The the code compiles because "util/Histogram.h" is included in "wsched/Scheduler.h". I suppose it is mildly evil in that including the header will make a small increase in compile time, and it appears "wsched/Scheduler.h" doesn't really need it either. I can do something about that.